File: rust/src/medium/yacht.rs
pub enum Category {
    Ones,
    Twos,
    Threes,
    Fours,
    Fives,
    Sixes,
    FullHouse,
    FourOfAKind,
    LittleStraight,
    BigStraight,
    Choice,
    Yacht,
}

type Dice = [u8; 5];
pub fn score(dice: Dice, category: Category) -> u8 {
    use Category::*;

    match category {
        Ones => count(1, dice),
        Twos => count(2, dice),
        Threes => count(3, dice),
        Fours => count(4, dice),
        Fives => count(5, dice),
        Sixes => count(6, dice),
        FullHouse => {
            let dices = (1..=6)
                .map(|x| (dice.iter().filter(|i| **i == x)).count())
                .filter(|x| *x == 2 || *x == 3)
                .sum();

            match dices {
                5 => dice.iter().sum(),
                _ => 0,
            }
        }
        FourOfAKind => {
            let kind = (1..=6).find(|x| (dice.iter().filter(|i| **i == *x)).count() >= 4);
            match kind {
                Some(kind) => dice.iter().filter(|i| **i == kind).take(4).sum(),
                None => 0,
            }
        }
        LittleStraight => {
            let mut dice = dice.to_vec();
            dice.sort_unstable();
            dice.dedup();

            match &dice[..] {
                [1, 2, 3, 4, 5] => 30,
                _ => 0,
            }
        }
        BigStraight => {
            let mut dice = dice.to_vec();
            dice.sort_unstable();
            dice.dedup();

            match &dice[..] {
                [2, 3, 4, 5, 6] => 30,
                _ => 0,
            }
        }
        Choice => dice.iter().sum(),
        Yacht => match dice.iter().all(|item| *item == dice[0]) {
            true => 50,
            false => 0,
        },
    }
}

fn count(number: u8, dice: Dice) -> u8 {
    number * dice.iter().filter(|x| **x == number).count() as u8
}

#[test]
fn test_yacht() {
    let expected = 50;

    assert_eq!(score([5, 5, 5, 5, 5], Category::Yacht), expected);
}

#[test]
fn test_four_of_a_kind() {
    let expected = 24;

    assert_eq!(score([6, 6, 4, 6, 6], Category::FourOfAKind), expected);
}

#[test]
fn test_four_of_a_kind_is_not_a_full_house() {
    let expected = 0;

    assert_eq!(score([1, 4, 4, 4, 4], Category::FullHouse), expected);
}

File: elixir/lib/medium/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(:ones, dice), do: Enum.count(dice, &(&1 == 1))
  def score(:twos, dice), do: Enum.count(dice, &(&1 == 2)) * 2
  def score(:threes, dice), do: Enum.count(dice, &(&1 == 3)) * 3
  def score(:fours, dice), do: Enum.count(dice, &(&1 == 4)) * 4
  def score(:fives, dice), do: Enum.count(dice, &(&1 == 5)) * 5
  def score(:sixes, dice), do: Enum.count(dice, &(&1 == 6)) * 6

  def score(:full_house, dice) do
    counts = dice |> counts() |> Map.values() |> Enum.sort()

    case counts do
      [2, 3] -> Enum.sum(dice)
      _ -> 0
    end
  end

  def score(:four_of_a_kind, dice) do
    kind = dice |> counts() |> Enum.find(fn {_key, val} -> val >= 4 end)

    case kind do
      {num, _count} -> num * 4
      nil -> 0
    end
  end

  def score(:little_straight, dice) do
    if Enum.sort(dice) == [1, 2, 3, 4, 5], do: 30, else: 0
  end

  def score(:big_straight, dice) do
    if Enum.sort(dice) == [2, 3, 4, 5, 6], do: 30, else: 0
  end

  def score(:choice, dice), do: Enum.sum(dice)

  def score(:yacht, [head | rest]) do
    if Enum.all?(rest, &(&1 == head)), do: 50, else: 0
  end

  defp counts(dice) do
    Enum.reduce(dice, %{}, fn num, counts ->
      Map.update(counts, num, 1, &(&1 + 1))
    end)
  end
end
